### PR TITLE
[Merged by Bors] - chore: Generalize results on `finrank` to rings.

### PR DESCRIPTION
**Pull request by erdOne**
A portion of results in `Mathlib/LinearAlgebra/FiniteDimensional.lean` were generalized to rings and moved to `Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean`. Most API lemmas for `FiniteDimensional` are kept but replaced with one lemma proofs. Definitions and niche lemmas are replaced by the generalized version completely.

---

Second attempt of #8872

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by riccardobrasca — COMMENTED** [review]
This looks like a great generalization, thanks a lot!!

Can you summarize the modification you did? Currently it is a little hard to follow everything (but it's clearly good).

**Review by riccardobrasca — COMMENTED** [review]
Since you are at it, can you remove or at least fix the line "The definition is marked as protected to avoid conflicts with _root_.rank" in `Mathlib.LinearAlgebra.Dimension`? We now have `[LinearMap.rank](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Dimension.html#LinearMap.rank)` so I think there is no need to protect this one (or maybe there is, but the doc is still imprecise at the moment).

File: Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean [hunk 1]
```diff
@@ -227,3 +231,488 @@ theorem Submodule.finrank_le_finrank_of_le {s t : Submodule R M} [Module.Finite
...
+@[simp]
+theorem FiniteDimensional.basisUnique_repr_eq_zero_iff [Module.Free R V] {ι : Type*} [Unique ι]
+    {h : finrank R V = 1} {v : V} {i : ι} :
+    (basisUnique ι h).repr v i = 0 ↔ v = 0 :=
+  ⟨fun hv =>
+    (basisUnique ι h).repr.map_eq_zero_iff.mp (Finsupp.ext fun j => Subsingleton.elim i j ▸ hv),
+    fun hv => by rw [hv, LinearEquiv.map_zero, Finsupp.zero_apply]⟩
+#align finite_dimensional.basis_unique.repr_eq_zero_iff FiniteDimensional.basisUnique_repr_eq_zero_iff
+
+theorem FiniteDimensional.cardinal_mk_le_finrank_of_linearIndependent [Module.Finite R V]
```
[the remark this answers is not in the document]
Comment (erdOne) [reply Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean:329]: I think I'll reorganize `LinearAlgebra/Dimension.lean`, `LinearAlgebra/Finrank.lean`, `LinearAlgebra/FreeModule/Finite/Rank.lean`, `LinearAlgebra/FreeModule/Rank.lean` (and maybe `LinearAlgebra.FiniteDimensional`?) into a new folder under `LinearAlgebra` after I generalized all the results, as there are already lots of misplaced lemmas and some files mentioned above also needs to be splitted.

**Comment by erdOne** [comment]
I updated the PR message.

File: Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean (hunk 1 above)
[the remark this answers is not in the document]
Comment (riccardobrasca) [reply Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean:329]: OK, no problem for doing it in another PR (anyway it is already a mess...). Can you just add a TODO?

**Comment by riccardobrasca** [comment]
I'll try to have a closer look tomorrow. In the meantime can you please fix the error?

File: Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean [hunk 2]
```diff
@@ -227,3 +232,496 @@ theorem Submodule.finrank_le_finrank_of_le {s t : Submodule R M} [Module.Finite
...
+
+end Submodule
+
+section
+
+variable [Ring R] [StrongRankCondition R] [AddCommGroup V] [Module R V]
+
+instance Module.Finite.finsupp {ι : Type*} [_root_.Finite ι] [Module.Finite R V] :
+    Module.Finite R (ι →₀ V) :=
+  Module.Finite.equiv (Finsupp.linearEquivFunOnFinite R V ι).symm
```
Comment (alreadydone) [inline Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean:725]: Can you just put these into RingTheory.Finiteness? They don't require StrongRankCondition at all.

**Review by alreadydone — COMMENTED** [review]
There are some minor issues but mostly LGTM!

File: Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean [hunk 3]
```diff
@@ -227,3 +232,496 @@ theorem Submodule.finrank_le_finrank_of_le {s t : Submodule R M} [Module.Finite
...
+  rw [Free.rank_eq_card_chooseBasisIndex, mk_lt_aleph0_iff]
+  exact ⟨fun h ↦ Finite.of_basis (Free.chooseBasis R V),
+    fun I ↦ Finite.of_fintype (Free.ChooseBasisIndex R V)⟩
+
+theorem FiniteDimensional.finrank_of_not_finite [Module.Free R V]
+    (h : ¬Module.Finite R V) :
+    FiniteDimensional.finrank R V = 0 :=
+  dif_neg (rank_lt_alpeh0_iff.not.mpr h)
+
+theorem Module.finite_of_finrank_pos (h : 0 < finrank R V) :
```
[the remark this answers is not in the document]
Comment (erdOne) [reply Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean:262]: I'm not planning to change the lemma statements in this PR.

File: Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean [hunk 4]
```diff
@@ -227,3 +232,496 @@ theorem Submodule.finrank_le_finrank_of_le {s t : Submodule R M} [Module.Finite
...
+    (Module.mk_finrank_eq_card_basis b).symm
+  -- porting note: replace `algebraMap.coe_one` with `Nat.cast_one`
+  simp only [Cardinal.eq_one_iff_unique, FiniteDimensional.finrank_self, Nat.cast_one] at A
+  exact Nonempty.some ((unique_iff_subsingleton_and_nonempty _).2 A)
+#align basis.unique Basis.unique
+
+variable (R V)
+
+/-- A finite rank free module has a basis indexed by `Fin (finrank R V)`. -/
+noncomputable def FiniteDimensional.finBasis [Module.Finite R V] :
```
[the remark this answers is not in the document]
Comment (erdOne) [reply Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean:297]: I'll have to rename them anyway. This has nothing to do with `FiniteDimensional`

File: Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean (hunk 2 above)
Comment (erdOne) [reply Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean:725]: A lot of lemmas in this file are already misplaced and will need a cleanup PR later. I can move these lemmas first if you insist.

File: Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean (hunk 4 above)
[the remark this answers is not in the document]
Comment (alreadydone) [reply Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean:296]: I just don't understand why you manually added the "wrong" namespace to each declaration. If you put them in a `namespace FiniteDimensional` section, wouldn't it be easier to change it to `namespace Module.Finite` in the future?

[the remark this answers is not in the document]
Comment (riccardobrasca) [reply Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean:296]: I agree with this comment. Is the namespace to keep the actual name? In this case we can add a TODO (but using `namespace` seems anyway a good idea).

File: Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean [hunk 5]
```diff
@@ -227,3 +232,496 @@ theorem Submodule.finrank_le_finrank_of_le {s t : Submodule R M} [Module.Finite
...
+open BigOperators
+
+open Finset
+
+/-- If a finset has cardinality larger than the rank of a module,
+then there is a nontrivial linear relation amongst its elements.
+-/
+theorem Module.exists_nontrivial_relation_of_finrank_lt_card
+    [Module.Finite R V] {t : Finset V}
+    (h : finrank R V < t.card) : ∃ f : V → R, ∑ e in t, f e • e = 0 ∧ ∃ x ∈ t, f x ≠ 0 := by
```
[the remark this answers is not in the document]
Comment (riccardobrasca) [reply Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean:486]: Add a TODO?

**Comment by riccardobrasca** [comment]
I think we can merge this (after the two small comments). It is surely not perfect, but perfect is the enemy of good, and this is a clear improvement wrt the current situation. Thanks!

bors d+

**Comment by erdOne** [comment]
bors merge

File: Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean (hunk 2 above)
Comment (alreadydone) [reply Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean:721]: As [mathlib docs shows]([Submodule.fg_iff_finite](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/FreeModule/Finite/Rank.html#Submodule.fg_iff_finite)) these already doesn't use StrongRankCondition. I removed it from `variable` in #9152 so that it won't be forgotten. I refrained from moving them, awaiting the misplacement cleanup PR.

File: Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean [hunk 6]
```diff
@@ -227,3 +232,492 @@ theorem Submodule.finrank_le_finrank_of_le {s t : Submodule R M} [Module.Finite
...
+theorem Module.finite_iff_of_rank_eq_nsmul {W} [AddCommGroup W]
+    [Module R W] [Module.Free R W] {n : ℕ} (hn : n ≠ 0)
+    (hVW : Module.rank R V = n • Module.rank R W) :
+    Module.Finite R V ↔ Module.Finite R W := by
+  simp only [← rank_lt_alpeh0_iff, hVW, nsmul_lt_aleph0_iff_of_ne_zero hn]
+
+/-- If a free module is of finite rank, then the cardinality of any basis is equal to its
+`finrank`. -/
+theorem Module.mk_finrank_eq_card_basis [Module.Finite R V]
+    {ι : Type w} (h : Basis ι R V) : (finrank R V : Cardinal.{w}) = #ι := by
```
Comment (alreadydone) [inline Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean:280]: Should this be `coe_finrank_eq_card_basis`? I guess we could also omit the `coe` ... `Cardinal.mk` doesn't appear in it.

File: Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean (hunk 5 above)
[the remark this answers is not in the document]
Comment (alreadydone) [reply Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean:490]: #9192
